test/common/http/codec_client_test.cc

```diff
@@ -128,6 +131,82 @@ TEST_F(CodecClientTest, DisconnectBeforeHeaders) {
+        inner_decoder = &decoder;
+        return inner_encoder;
+      }));
+
+  Http::MockStreamDecoder outer_decoder;
+  Http::StreamEncoder& request_encoder = client_->newStream(outer_decoder);
+  Http::MockStreamCallbacks callbacks;
+  request_encoder.getStream().addCallbacks(callbacks);
+  connection_cb_->onEvent(Network::ConnectionEvent::Connected);
+
+  Http::HeaderMapPtr response_headers{new TestHeaderMapImpl{{":status", "200"}}};
+  EXPECT_CALL(outer_decoder, decodeHeaders_(Pointee(Ref(*response_headers)), false));
+  inner_decoder->decodeHeaders(std::move(response_headers), false);
+
+  Buffer::OwnedImpl buffer("hello");
+  EXPECT_CALL(outer_decoder, decodeData(Ref(buffer), true));
+  inner_decoder->decodeData(buffer, true);
+  EXPECT_NE(client_->idleTimer(), nullptr);
+
+  // Close the client and validate idleTimer is reset
+  EXPECT_EQ(client_->numActiveRequests(), 0);
+  client_->close();
+  connection_cb_->onEvent(Network::ConnectionEvent::LocalClose);
+  EXPECT_EQ(client_->idleTimer(), nullptr);
+}
+
+TEST_F(CodecClientTest, IdleTimerClientRemoteCloseWithActiveRequests) {
+  Http::StreamDecoder* inner_decoder;
+  NiceMock<Http::MockStreamEncoder> inner_encoder;
+  EXPECT_CALL(*codec_, newStream(_))
+      .WillOnce(Invoke([&](Http::StreamDecoder& decoder) -> Http::StreamEncoder& {
+        inner_decoder = &decoder;
+        return inner_encoder;
+      }));
+
+  Http::MockStreamDecoder outer_decoder;
+  Http::StreamEncoder& request_encoder = client_->newStream(outer_decoder);
+  Http::MockStreamCallbacks callbacks;
+  request_encoder.getStream().addCallbacks(callbacks);
+
+  // When we get a remote close with an active request validate idleTimer is reset after client
+  // close
+  EXPECT_CALL(callbacks, onResetStream(StreamResetReason::ConnectionTermination));
+  EXPECT_CALL(*codec_, dispatch(_));
+  EXPECT_NE(client_->numActiveRequests(), 0);
+  connection_cb_->onEvent(Network::ConnectionEvent::Connected);
+  connection_cb_->onEvent(Network::ConnectionEvent::RemoteClose);
+  EXPECT_EQ(client_->idleTimer(), nullptr);
+}
+
+TEST_F(CodecClientTest, IdleTimerClientLocalCloseWithActiveRequests) {
+  Http::StreamDecoder* inner_decoder;
+  NiceMock<Http::MockStreamEncoder> inner_encoder;
+  EXPECT_CALL(*codec_, newStream(_))
+      .WillOnce(Invoke([&](Http::StreamDecoder& decoder) -> Http::StreamEncoder& {
+        inner_decoder = &decoder;
+        return inner_encoder;
+      }));
+
+  Http::MockStreamDecoder outer_decoder;
+  Http::StreamEncoder& request_encoder = client_->newStream(outer_decoder);
+  Http::MockStreamCallbacks callbacks;
+  request_encoder.getStream().addCallbacks(callbacks);
+
+  // When we get a local close with an active request validate idleTimer is reset after client close
+  EXPECT_CALL(callbacks, onResetStream(StreamResetReason::ConnectionTermination));
+  connection_cb_->onEvent(Network::ConnectionEvent::Connected);
+  connection_cb_->onEvent(Network::ConnectionEvent::LocalClose);
```

Comment:
Same here with TODO. Optimally, we should also call close() on the client to make the test more realistic.